compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12804,6 +12804,220 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+ * \endverbatim
+ *
+ * Also note that for unsigned comparison types other than equality and inequality,
+ * add/sub simplifications are not possible due to how values are interpreted
+ * in the current IL nodes. Essentially, the value of an iconst node is neither
+ * signed nor unsigned; and its parent is free to interpret its value.
+ *
+ * For example, the following transformation is incorrect:
+ *
+ * \varbatim
+ *
+ * Before:
+ *
+ * ifiucmplt --> goto ...
+ *     isub (cannotOverflow)
+ *         iload x
+ *         iconst -1
+ *     iconst 100
+ *
+ * After:
+ *
+ * ifiucmplt --> goto ...
+ *     iload x
+ *     iconst 99
+ *
+ * \endverbatim
+ *
+ * If x were -1, the first tree would evaluate to true since it's comparing if zero is less than 100.
+ * But in the second tree, ifiucmplt interprets the value of x as UINT_MAX. Thus,
+ * the second tree is comparing UINT_MAX with 99; and the whole comparison node evalutes
+ * to false.
+ *
+ * Unsigned equality and inequality comparisons are not sensitive to sign interpretation problems.
+ *
+ * \param node          the comparison node being simpilfied
+ * \param s             the simplifier object
+*/
+TR::Node* removeArithmeticsUnderIntegralCompare(TR::Node* node,
+                                                TR::Simplifier * s)
+   {
+   if (s->comp()->getOption(TR_DisableIntegerCompareSimplification))
+      return node;
+
+   TR_ASSERT(node->getOpCode().isBooleanCompare()
+             && node->getNumChildren() > 0
+             && node->getFirstChild()->getOpCode().isInteger(),
+             "Unsupported opCode data type when removing integer arithmetics for integer comparison nodes.");
+
+   TR::Node* opNode      = node->getFirstChild();
+   TR::Node* secondChild = node->getSecondChild();
+   TR::ILOpCodes op      = opNode->getOpCodeValue();
+
+   // Get recognized operation types. Unsigned IL OpCodes are deprecated and not supported here.
+   bool isAddOp = (op == TR::iadd || op == TR::ladd || op == TR::sadd || op == TR::badd);
+   bool isSubOp = (op == TR::isub || op == TR::lsub || op == TR::ssub || op == TR::bsub);
+
+   bool isUnsignedCompare = node->getOpCode().isUnsignedCompare();
+   bool isEqOrNotEqCompare = node->getOpCode().isCompareForEquality();
+
+   bool isSignednessTypeSupported = isEqOrNotEqCompare || (!isUnsignedCompare && opNode->chkCannotOverflow());
+
+   if ((isAddOp || isSubOp)
+         && isSignednessTypeSupported
+         && opNode->getSecondChild()->getOpCode().isLoadConst()
+         && secondChild->getOpCode().isLoadConst()
+         // Only perform the simplification on the first occurance of the add/sub node
+         && (opNode->getFutureUseCount() == opNode->getReferenceCount() - 1))
+      {
+      int64_t oldConst1 = 0, oldConst2 = 0, newConst = 0;     // for signed compare types
+      int64_t signedMax = 0, signedMin = 0;
+
+      uint64_t oldUConst1 = 0, oldUConst2 = 0, newUConst = 0;
```

Comment:
The two sets of constants here are redundant, and they cause a lot of repeated code just below. Could we use `uint64_t` only?